PWGUD/Tasks/upcMft.cxx

```diff
@@ -0,0 +1,313 @@
+    LOG(info) << "Setting magnetic field to current " << grpmag->getL3Current()
+              << " A for run " << bc.runNumber()
+              << " from its GRPMagField CCDB object";
+    o2::base::Propagator::initFieldFromGRP(grpmag); // for some reason this is necessary for the next next line
+    runNumber = bc.runNumber();
+
+    o2::field::MagneticField* field = static_cast<o2::field::MagneticField*>(TGeoGlobalMagField::Instance()->GetField());
+
+    Bz = field->getBz(centerMFT); // gives error if the propagator is not initFielded
+    LOG(info) << "The field at the center of the MFT is Bz = " << Bz;
+  }
+
+  void process(aod::BCs const& bcs,
+               aod::Collisions const& collisions,
+               aod::MFTTracks const& tracksMFT,
+               aod::AmbiguousMFTTracks const& ambiguousTracksMFT,
+               aod::FT0s const& ft0s,
+               ExtBCs const& ebcs)
+  {
+
+    std::map<int, BcTracks> bcGroups; // Map of tracks that belong to a given bunch crossing. One track can be associated with multiple BCs.
+
+    initCCDB(ebcs.begin()); // Bz is needed for propagation
+
+    for (auto& ambiguousTrackMFT : ambiguousTracksMFT) {
+      auto mftBCs = ambiguousTrackMFT.bc();
+      for (auto& mftBC : mftBCs) {
+        int bcId = mftBC.globalBC();        // Get global BC that will be used for grouping.
+        auto element = bcGroups.find(bcId); // Find the iterator poiniting to value associated with this BCid.
+        if (element == bcGroups.end()) {    // If this BCid was encountered for the first time, create a vector containing the current track
+          BcTracks bct;
+          bct.tracks.push_back(ambiguousTrackMFT); // was ambiguousTrackMFT
+
+          auto track = ambiguousTrackMFT.mfttrack();
+          std::vector<double> v1; // Temporary null vector for the computation of the covariance matrix
+          SMatrix55 tcovs(v1.begin(), v1.end());
+          SMatrix5 tpars(track.x(), track.y(), track.phi(), track.tgl(), track.signed1Pt());
+          o2::track::TrackParCovFwd trackPar{track.z(), tpars, tcovs, track.chi2()};
+          bct.trackParams.push_back(trackPar);
+          bcGroups.emplace(std::make_pair(bcId, bct));
+        } else { // If this BcId was already encountered, add the track to the existing vector
+          element->second.tracks.push_back(ambiguousTrackMFT);
+          auto track = ambiguousTrackMFT.mfttrack();
+          std::vector<double> v1; // Temporary null vector for the computation of the covariance matrix
+          SMatrix55 tcovs(v1.begin(), v1.end());
+          SMatrix5 tpars(track.x(), track.y(), track.phi(), track.tgl(), track.signed1Pt());
+          o2::track::TrackParCovFwd trackPar{track.z(), tpars, tcovs, track.chi2()};
+          element->second.trackParams.push_back(trackPar);
+        }
+      }
+
+      for (auto& ft0 : ft0s) {
+        int collisionBC = (bcs.iteratorAt(ft0.bcId())).globalBC(); // Get global BC that will be used for grouping.
```

Comment:
You could also use ft0.bc() to get the related BC, instead of bcs.iteratorAt(ft0.bcId())